crates/turborepo-filewatch/src/hash_watcher.rs

```diff
@@ -0,0 +1,847 @@
+
+pub struct HashWatcher {
+    _exit_tx: oneshot::Sender<()>,
+    _handle: tokio::task::JoinHandle<()>,
+    query_tx: mpsc::Sender<Query>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct HashSpec {
+    pub package_path: AnchoredSystemPathBuf,
+    pub inputs: Option<GlobSet>,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("package hashing encountered an error: {0}")]
+    HashingError(String),
+    #[error("file hashing is not available: {0}")]
+    Unavailable(String),
+    #[error("package not found: {} {:?}", .0.package_path, .0.inputs)]
+    UnknownPackage(HashSpec),
+}
+
+// Communication errors that all funnel to Unavailable
+
+impl From<watch::error::RecvError> for Error {
+    fn from(e: watch::error::RecvError) -> Self {
+        Self::Unavailable(e.to_string())
+    }
+}
+
+impl From<oneshot::error::RecvError> for Error {
+    fn from(e: oneshot::error::RecvError) -> Self {
+        Self::Unavailable(e.to_string())
+    }
+}
+
+impl<T> From<mpsc::error::SendError<T>> for Error {
+    fn from(e: mpsc::error::SendError<T>) -> Self {
+        Self::Unavailable(e.to_string())
+    }
+}
+
+impl HashWatcher {
+    pub fn new(
+        repo_root: AbsoluteSystemPathBuf,
+        package_discovery: watch::Receiver<Option<DiscoveryData>>,
+        file_events: OptionalWatch<broadcast::Receiver<Result<Event, NotifyError>>>,
+        scm: SCM,
+    ) -> Self {
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let (query_tx, query_rx) = mpsc::channel(16);
+        let subscriber = Subscriber::new(repo_root, package_discovery, scm, query_rx);
+        let handle = tokio::spawn(subscriber.watch(exit_rx, file_events));
+        Self {
+            _exit_tx: exit_tx,
+            _handle: handle,
+            query_tx,
+        }
+    }
+
+    // Note that this does not wait for any sort of ready signal. The watching
+    // process won't respond until filewatching is ready, but there is no
+    // guarantee that package data or file hashing will be done before
+    // responding. Both package discovery and file hashing can fail depending on the
+    // state of the filesystem, so clients will need to be robust to receiving
+    // errors.
+    pub async fn get_file_hashes(&self, hash_spec: HashSpec) -> Result<GitHashes, Error> {
+        let (tx, rx) = oneshot::channel();
+        self.query_tx.send(Query::GetHash(hash_spec, tx)).await?;
+        rx.await?
+    }
+}
+
+struct Subscriber {
+    repo_root: AbsoluteSystemPathBuf,
+    package_discovery: watch::Receiver<Option<DiscoveryData>>,
+    query_rx: mpsc::Receiver<Query>,
+    scm: SCM,
+}
+
+enum Query {
+    GetHash(HashSpec, oneshot::Sender<Result<GitHashes, Error>>),
+}
+
+// Version is a type that exists to stamp an asynchronous hash computation with
+// a version so that we can ignore completion of outdated hash computations.
+#[derive(Clone, Default)]
+struct Version(Arc<()>);
```

Comment:
I'm curious, is this a pattern that you've seen before in Rust? I *think* it's fine, but I could see a particularly zealous compiler optimizing zero sized types wrapped in Arcs to point to the same constant chunk of memory.

Reply:
I verified it in a test before adding it, but I doubt there were aggressive optimizations applied. I'll see if I can dig up any opinions on how safe it is.

Reply:
Opted to switch an `AtomicUsize` instead.